zero_bin/ops/src/lib.rs

```diff
@@ -0,0 +1,307 @@
+        let input = all_data.0.clone();
+        let _span = TxProofSpan::new(&input);
+        let proof = if self.save_inputs_on_error {
+            zero_bin_common::prover_state::p_manager()
+                .generate_segment_proof(all_data)
+                .map_err(|err| {
+                    if let Err(write_err) = save_inputs_to_disk(
+                        format!(
+                            "b{}_txn_{}_input.log",
+                            input.block_metadata.block_number, input.txn_number_before
+                        ),
+                        input,
+                    ) {
+                        error!("Failed to save txn proof input to disk: {:?}", write_err);
+                    }
+
+                    FatalError::from_anyhow(err, FatalStrategy::Terminate)
+                })?
+        } else {
+            zero_bin_common::prover_state::p_manager()
+                .generate_segment_proof(all_data)
+                .map_err(|err| FatalError::from_anyhow(err, FatalStrategy::Terminate))?
+        };
+
+        Ok(proof.into())
+    }
+}
+
+#[cfg(feature = "test_only")]
+impl Operation for SegmentProof {
+    type Input = AllData;
+    type Output = ();
+
+    fn execute(&self, input: Self::Input) -> Result<Self::Output> {
+        let gen_input = input.0;
+        let _span = TxProofSpan::new(&gen_input);
+
+        if self.save_inputs_on_error {
+            evm_arithmetization::prover::testing::simulate_execution::<proof_gen::types::Field>(
+                gen_input.clone(),
+            )
+            .map_err(|err| {
+                if let Err(write_err) = save_inputs_to_disk(
+                    format!(
+                        "b{}_txn_{}_input.log",
+                        gen_input.block_metadata.block_number, gen_input.txn_number_before
+                    ),
+                    gen_input,
+                ) {
+                    error!("Failed to save txn proof input to disk: {:?}", write_err);
+                }
+
+                FatalError::from_anyhow(err, FatalStrategy::Terminate)
+            })?;
+        } else {
+            evm_arithmetization::prover::testing::simulate_execution::<proof_gen::types::Field>(
+                gen_input.clone(),
+            )
+            .map_err(|err| FatalError::from_anyhow(err, FatalStrategy::Terminate))?;
+        }
+
+        Ok(())
+    }
+}
+
+/// RAII struct to measure the time taken by a transaction proof.
+///
+/// - When created, it starts a span with the transaction proof id.
+/// - When dropped, it logs the time taken by the transaction proof.
+struct TxProofSpan {
+    _span: tracing::span::EnteredSpan,
+    start: Instant,
+    descriptor: String,
+}
+
+impl TxProofSpan {
+    /// Get a unique id for the transaction proof.
+    fn get_id(ir: &GenerationInputs) -> String {
+        format!(
+            "b{} - {}",
+            ir.block_metadata.block_number, ir.txn_number_before
+        )
+    }
```

Comment:
This should be a `SegmentProofSpan` now, and the ID should contain the txn nb before, txn nb after (because of batching), and the segment we're in (otherwise we won't be able to distinguish all the duplicates).